src/gui/advancedsettings.cpp

```diff
@@ -508,10 +511,21 @@ void AdvancedSettings::loadAdvancedSettings()
     addRow(DISK_IO_TYPE, tr("Disk IO type (requires restart)") + u' ' + makeLink(u"https://www.libtorrent.org/single-page-ref.html#default-disk-io-constructor", u"(?)")
            , &m_comboBoxDiskIOType);
 #endif
-    // Enable OS cache
-    m_checkBoxOsCache.setChecked(session->useOSCache());
-    addRow(OS_CACHE, (tr("Enable OS cache") + u' ' + makeLink(u"https://www.libtorrent.org/reference-Settings.html#disk_io_write_mode", u"(?)"))
-            , &m_checkBoxOsCache);
+    // Disk IO read mode
+    m_comboBoxDiskIOReadMode.addItem(tr("Disable OS cache"), QVariant::fromValue(BitTorrent::DiskIOReadMode::DisableOSCache));
+    m_comboBoxDiskIOReadMode.addItem(tr("Enable OS cache"), QVariant::fromValue(BitTorrent::DiskIOReadMode::EnableOSCache));
+    m_comboBoxDiskIOReadMode.setCurrentIndex(m_comboBoxDiskIOReadMode.findData(QVariant::fromValue(session->diskIOReadMode())));
+    addRow(DISK_IO_READ_MODE, (tr("Disk IO read mode") + u' ' + makeLink(u"https://www.libtorrent.org/reference-Settings.html#disk_io_read_mode", u"(?)"))
+            , &m_comboBoxDiskIOReadMode);
+    // Disk IO write mode
+    m_comboBoxDiskIOWriteMode.addItem(tr("Disable OS cache"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::DisableOSCache));
+    m_comboBoxDiskIOWriteMode.addItem(tr("Enable OS cache"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::EnableOSCache));
+#ifdef QBT_USES_LIBTORRENT2
+    m_comboBoxDiskIOWriteMode.addItem(tr("Write-through"), QVariant::fromValue(BitTorrent::DiskIOWriteMode::WriteThrough));
+#endif
+    m_comboBoxDiskIOWriteMode.setCurrentIndex(m_comboBoxDiskIOWriteMode.findData(QVariant::fromValue(session->diskIOWriteMode())));
```

Comment:
What is selected if `session->diskIOWriteMode()` returns write-through when there is librorrent-1.2 build used?

Reply:
>What is selected if session->diskIOWriteMode() returns write-through when there is librorrent-1.2 build used?

Now `session::diskIOWriteMode()` will never return `write-through` when using librorrent 1.2. It will return the default value.

You can try with the Windows artifacts produced by github actions.

Reply:
>I can't figure out exactly where it discards invalid values, so it looks unreliable to me.

`std::is_enum_v<T>` case in `SettingsStorage::loadValue()` will detect invalid values and switch to default value.

Reply:
> > I can't figure out exactly where it discards invalid values, so it looks unreliable to me.
> 
> `std::is_enum_v<T>` case in `SettingsStorage::loadValue()` will detect invalid values and switch to default value.

But CachedSettingValue don't use it after first load. So if you set invalid value you'll get the same invalid value next time, no?

Reply:
> Don't know what you mean.

https://github.com/qbittorrent/qBittorrent/blob/e2b140b8ddb315ecc2ece55ce3dd54a3ea12817f/src/base/settingvalue.h#L95-L103

Let's imagine that `value` is invalid. Now `m_cache` has invalid value which is returned by `get()`:

https://github.com/qbittorrent/qBittorrent/blob/e2b140b8ddb315ecc2ece55ce3dd54a3ea12817f/src/base/settingvalue.h#L85-L88

Reply:
>Let's imagine that value is invalid. Now m_cache has invalid value which is returned by get():

OK, please refer to my second paragraph of my last reply.
I don't really ack this is a problem nor intend to solve this problem in this PR.

Reply:
> You can't set it to invalid value at run time (with libtorrent 1.2)

Here:
https://github.com/qbittorrent/qBittorrent/blob/5b730ffb72b3e3f70cd54ac56707d3830e809809/src/webui/api/appcontroller.cpp#L826-L827

Reply:
> > You can't set it to invalid value at run time (with libtorrent 1.2)
> 
> Here:

I've voiced explicitly in the past that static_cast to enum values won't work reliably and nobody really understands it. And now here we are...

I'm not going to resolve it in this PR and read the rest of the [second paragraph](https://github.com/qbittorrent/qBittorrent/pull/17367#discussion_r921101035).

Reply:
>that hasn't gotten much better in the latest standards.

It may never be resolved... just curious, which other language has better enum types?

>Well, this is really a broader problem.

Yes. And after some thinking it seems possible to patch it (in a another PR of course): modify *SettingValue and SettingsStorage classes and therefore SettingValue & SettingsStorage will be able to reject/ignore invalid enum values on assignment. But do you think it is worth it?

Reply:
> > that hasn't gotten much better in the latest standards.
> 
> It may never be resolved...

This will **most likely** not be resolved. I hoped for this when C++ began to develop after a long stagnation. Now I don't even hope.
>just curious, which other language has better enum types?

I can't answer you exactly. In my youth, I got acquainted with several languages... I'm not even sure that the features of enumerations from several languages just didn't mix in my memory. But if I'm not mistaken, in some of them enumeration was really a limited set of different named values with integrity control at the language level (i.e. a variable of enumeration type cannot contain an invalid value).

Reply:
> And after some thinking it seems possible to patch it (in a another PR of course): modify *SettingValue and SettingsStorage classes and therefore SettingValue & SettingsStorage will be able to reject/ignore invalid enum values on assignment. But do you think it is worth it?

I don't think changes at this level will do much good other than making SettingValue and CachedSettingValue behave consistently with enums.

Reply:
> > > that hasn't gotten much better in the latest standards.
> > 
> > 
> > It may never be resolved...
> 
> This will **most likely** not be resolved. I hoped for this when C++ began to develop after a long stagnation. Now I don't even hope.

I even often thought about doing some kind of custom implementation of enumerations, but unfortunately it can't be done without using the macros of the C++ preprocessor (or some third-party preprocessor), and I don't like it, because the code doesn't look native. In addition, there are several existing implementations of "smart" enumerations, so we could take some of them if we wanted to (e.g. https://github.com/seleznevae/smart_enum).